kpatch: retry kpatch load on failure

jpoimboe opened the pull request:
If a kpatch load fails, try again, up to 5 times.

It will usually only fail if the activeness safety check failed, in
which case trying again might work.  Unfortunately, insmod doesn't
return the error reported by the module, so we can't distinguish
activeness safety errors from other types of errors.

flaming-toast commented on kpatch/kpatch:
```diff
@@ -163,7 +163,16 @@ load_module () {
+		fi
+		insmod "$1" "$2" && break
+	done
+	[[ $i = 5 ]] && die "failed to load module $modname"
```
If `insmod` happens to succeed on the 5th try (i = 5), then we will `die` here. Maybe set a fail flag within the loop (at i=5, if insmod fails, then set fail=1...) and then check it outside of the for loop?

joe-lawrence replied:
Alternatively move the conditional check at the bottom of the loop.  Any successful ```insmod``` will break the loop and skip the check...  Or do something like: ```insmod "$1" "$2" && loaded=1 && break``` to employ a success flag without having duplicate "5" constants laying about.

flaming-toast replied:
@joe-lawrence Ah, I like your idea better than mine :-) (ditching the conditional and moving the sleep and echo after the insmod)

jpoimboe replied:
Will fix.

flaming-toast commented on kpatch/kpatch:
```diff
@@ -163,7 +163,16 @@ load_module () {
+			echo "retrying..."
+			sleep 2
+		fi
+		insmod "$1" "$2" && break
```
Do we still need to call `insmod` with `$2`? Looks like nothing in the script sets it anymore. It used to be used to pass the `replace` param, but IIRC the `kpatch replace` option was deprecated in commit 8e8de4718d8

jpoimboe replied:
Oops, yeah.  This is some ugly code.  These functions should really be using local variables to deal with the function arguments.

jpoimboe commented:
Updated.

- rewrote loop based on all the comments
- cleaned up load_module() to use local variables
- removed unused 2nd argument to insmod
- added specific checking for insmod error, retry only makes sense for EBUSY.

joe-lawrence commented on kpatch/kpatch:
```diff
@@ -165,7 +165,27 @@ load_module () {
+		out=$(insmod "$module" 2>&1)
+		[[ -z $out ]] && break
+		echo $out 1>&2
+		[[ ! $out =~ "Device or resource busy" ]] &&
```
I have no i18n experience, but is it assured that insmod output will be in English?  (Too bad it doesn't propagate out the error code as insmod exit status.)

flaming-toast replied:
@joe-lawrence That's an interesting question. I looked at the insmod sources from kmod and old module-init-tools sources (superceded by kmod) and they both have always used `strerror` to print the message, *not* `strerror_l`,  which would print the locale-dependent/translated version of the error message. So I think the message will always be in English.

(For the curious, the translated glibc error messages used by `strerror_l` and others can be found here: https://sourceware.org/git/?p=glibc.git;a=tree;f=po;h=b6fb91b5b5b9730f24a0960bf0b0db766fc54c17;hb=HEAD)

joe-lawrence commented on kpatch/kpatch:
```diff
@@ -165,7 +165,27 @@ load_module () {
+		# "Device or resource busy" means the activeness safety check
+		# failed.  Retry in a few seconds.
+		i=$((i+1))
+		if [[ $i = 5 ]]; then
```
I always get tripped up with shell scripts... is ```=``` legit here, or do you need ```-eq``` for numerical test?

jpoimboe replied:
`=` is legit because all variables are really strings in bash.  So it's testing for the string "5".  `-eq` would have worked too.

joe-lawrence replied:
Okay, like I said, it just gets confusing, especially with ```-gt``` vs ```>```.  I did test this in a shell and you're right, the string comparison will match.
